linalg/ginkgo.cpp

```diff
@@ -28,93 +28,293 @@ namespace mfem
-      executor = gko::CudaExecutor::create(0, gko::OmpExecutor::create());
+      imp_rel_criterion  = ImplicitResidualCriterionFactory::build()
+                           .with_reduction_factor(sqrt(rel_tol))
+                           .with_baseline(gko::stop::mode::initial_resnorm)
+                           .on(executor);
+      imp_abs_criterion  = ImplicitResidualCriterionFactory::build()
+                           .with_reduction_factor(sqrt(abs_tol))
+                           .with_baseline(gko::stop::mode::absolute)
+                           .on(executor);
+      combined_factory =
+         gko::stop::Combined::build()
+         .with_criteria(imp_rel_criterion,
+                        imp_abs_criterion,
+                        gko::stop::Iteration::build()
+                        .with_max_iters(static_cast<unsigned long>(max_iter))
+                        .on(executor))
+         .on(executor);
    }
    else
    {
-      mfem::err <<
-                " exec_type needs to be one of the three strings: \"reference\", \"cuda\" or \"omp\" "
-                << std::endl;
-   }
-   using ResidualCriterionFactory = gko::stop::ResidualNormReduction<>;
-   residual_criterion             = ResidualCriterionFactory::build()
-                                    .with_reduction_factor(rel_tol)
-                                    .on(executor);
-
-   combined_factory =
-      gko::stop::Combined::build()
-      .with_criteria(residual_criterion,
-                     gko::stop::Iteration::build()
-                     .with_max_iters(max_iter)
-                     .on(executor))
-      .on(executor);
+      rel_criterion  = ResidualCriterionFactory::build()
+                       .with_reduction_factor(rel_tol)
+                       .with_baseline(gko::stop::mode::initial_resnorm)
+                       .on(executor);
+      abs_criterion  = ResidualCriterionFactory::build()
+                       .with_reduction_factor(abs_tol)
+                       .with_baseline(gko::stop::mode::absolute)
+                       .on(executor);
+      combined_factory =
+         gko::stop::Combined::build()
+         .with_criteria(rel_criterion,
+                        abs_criterion,
+                        gko::stop::Iteration::build()
+                        .with_max_iters(static_cast<unsigned long>(max_iter))
+                        .on(executor))
+         .on(executor);
+   }
+
+   needs_wrapped_vecs = false;
+   sub_op_needs_wrapped_vecs = false;
 }
 
 void
-GinkgoIterativeSolverBase::initialize_ginkgo_log(gko::matrix::Dense<double>* b)
+GinkgoIterativeSolver::initialize_ginkgo_log(gko::matrix::Dense<double>* b)
+const
 {
    // Add the logger object. See the different masks available in Ginkgo's
    // documentation
    convergence_logger = gko::log::Convergence<>::create(
                            executor, gko::log::Logger::criterion_check_completed_mask);
    residual_logger = std::make_shared<ResidualLogger<>>(executor,
```

Comment:
I'm not seeing warnings with my current test setup, but I know I've seen them before with other compilers.  Unfortunately, I think I remember that they come from the Ginkgo class itself (`Convergence`).  I'll try some more compilers and see if I can reproduce again.

Reply:
I see the warnings with Apple clang version 12.0.0. If they come from Ginkgo itself then there is not much we can do in this PR, but it would be nice to address it in Ginkgo if possible.